shell/platform/darwin/macos/framework/Source/FlutterExternalTextureGL.mm

```diff
@@ -0,0 +1,91 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "flutter/shell/platform/darwin/macos/framework/Source/FlutterExternalTextureGL.h"
+
+#import <AppKit/AppKit.h>
+#import <CoreVideo/CoreVideo.h>
+#import <OpenGL/gl.h>
+
+static void OnGLTextureRelease(CVPixelBufferRef pixelBuffer) {
+  CVPixelBufferRelease(pixelBuffer);
+}
+
+@implementation FlutterExternalTextureGL {
+  /**
+   * OpenGL texture cache.
+   */
+  CVOpenGLTextureCacheRef _openGLTextureCache;
+  /**
+   * The pixel buffer copied from the user side will be released
+   * when the flutter engine renders it.
+   */
+  CVPixelBufferRef _pixelBuffer;
+  /**
+   * User side texture object, used to copy pixel buffer.
+   */
+  id<FlutterTexture> _texture;
+}
+
+- (instancetype)initWithFlutterTexture:(id<FlutterTexture>)texture {
+  self = [super init];
+  if (self) {
+    _texture = texture;
+  }
+  return self;
+}
+
+- (int64_t)textureID {
+  return reinterpret_cast<int64_t>(self);
+}
+
+- (BOOL)populateTextureWithWidth:(size_t)width
+                          height:(size_t)height
+                   openGLTexture:(FlutterOpenGLTexture*)openGLTexture {
+  // Copy the pixel buffer from the FlutterTexture instance implemented on the user side.
+  _pixelBuffer = [_texture copyPixelBuffer];
+
+  if (!_pixelBuffer) {
+    return NO;
+  }
+
+  // Create the opengl texture cache if necessary.
+  if (!_openGLTextureCache) {
+    CGLContextObj context = [NSOpenGLContext currentContext].CGLContextObj;
+    CGLPixelFormatObj format = CGLGetPixelFormat(context);
+    if (CVOpenGLTextureCacheCreate(kCFAllocatorDefault, NULL, context, format, NULL,
+                                   &_openGLTextureCache) != kCVReturnSuccess) {
+      NSLog(@"Could not create texture cache.");
+      CVPixelBufferRelease(_pixelBuffer);
+      return NO;
+    }
+  }
+
+  // Try to clear the cache of OpenGL textures to save memory.
+  CVOpenGLTextureCacheFlush(_openGLTextureCache, 0);
+
+  CVOpenGLTextureRef cvOpenGLTexture = NULL;
+  if (CVOpenGLTextureCacheCreateTextureFromImage(kCFAllocatorDefault, _openGLTextureCache,
+                                                 _pixelBuffer, NULL,
+                                                 &cvOpenGLTexture) != kCVReturnSuccess) {
+    CVPixelBufferRelease(_pixelBuffer);
+    return NO;
+  }
+  CVPixelBufferRelease(_pixelBuffer);
```

Comment:
I'm very confused by the handling of `_pixelBuffer` in this code. It's stored as an ivar, but always assigned at the beginning of this method and never read in other methods, so I don't understand why it's not a local variable. Then you're releasing it here, so the stored value is, unless I'm missing something, a dangling pointer. You're also still using it later in this method after releasing it.

It seems like either there something very non-obvious about how this object is managed (in which case, please add explanatory comments), or it should just be a local (and released at the end of the method, not while you are still using it).

Reply:
Because in copyPixelBuffer, the reference count is incremented to prevent early release, like this:

```objc
- (CVPixelBufferRef)copyPixelBuffer {
     if(_videoPixelBufferRef != nil){
         CVBufferRetain(_videoPixelBufferRef);
         return _videoPixelBufferRef;
     }
}
```

So after the CVOpenGLTextureCacheCreateTextureFromImage copy process, we need to decrement the object reference.

Reply:
I understand why there are releases to balance the copy; that wasn't what I asked. Your comment doesn't seem to address my questions. Again, they are:
- Why is _pixelBuffer an ivar rather than a local variable?
- If it's going to be an ivar, why are you releasing it without nulling it?
- Why are you using it after releasing it?

Reply:
Ok, just use local variable now.